sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp

```diff
@@ -0,0 +1,452 @@
+    joint_matrix_hip<
+        S, sycl::ext::oneapi::experimental::matrix::use::accumulator, M, N,
+        sycl::ext::oneapi::experimental::matrix::layout::dynamic> &res,
+    multi_ptr<T, Space, IsDecorated> src, size_t stride, Group &sg) {
+  const auto idx = sg.get_group_linear_id() * sg.get_local_range()[0] +
+                   sg.get_local_linear_id();
+
+  if constexpr (std::is_same_v<S, double>) {
+    const auto thread_x = idx % N;
+    const auto thread_y = idx / N;
+
+    if constexpr (Layout ==
+                  sycl::ext::oneapi::experimental::matrix::layout::row_major) {
+      for (int i = 0; i < 4; ++i) {
+        const int s_idx = thread_x + i * 4 * stride + thread_y * stride;
+        res.data[i] = src[s_idx];
+      }
+    } else {
+      for (int i = 0; i < 4; ++i) {
+        const int s_idx = i * 4 + thread_x * stride + thread_y;
+        res.data[i] = src[s_idx];
+      }
+    }
+  } else if constexpr (std::is_same_v<S, float> || std::is_same_v<S, int32_t>) {
+    if constexpr (M == 16 && N == 16) {
+      const auto thread_x = idx % N;
+      const auto thread_y = idx / N;
+
+      if constexpr (Layout == sycl::ext::oneapi::experimental::matrix::layout::
+                                  row_major) {
+        for (int i = 0; i < 4; ++i) {
+          const int s_idx = thread_x + i * stride + thread_y * 4 * stride;
+          res.data[i] = src[s_idx];
+        }
+      } else {
+        for (int i = 0; i < 4; ++i) {
+          const int s_idx = i + thread_x * stride + thread_y * 4;
+          res.data[i] = src[s_idx];
+        }
+      }
+    } else if constexpr (M == 32 && N == 32) {
+      const auto thread_x = idx % N;
+      const auto thread_y = idx / N;
+
+      if constexpr (Layout == sycl::ext::oneapi::experimental::matrix::layout::
+                                  row_major) {
+        for (int j = 0; j < 4; ++j) {
+          for (int i = 0; i < 4; ++i) {
+            const int s_idx =
+                thread_x + i * stride + thread_y * 4 * stride + j * 8 * N;
+            res.data[i + 4 * j] = src[s_idx];
+          }
+        }
+      } else {
+        for (int j = 0; j < 4; ++j) {
+          for (int i = 0; i < 4; ++i) {
+            const int s_idx = i + thread_x * stride + thread_y * 4 + j * 8;
+            res.data[i + 4 * j] = src[s_idx];
+          }
+        }
+      }
+    }
+  }
+}
+
+template <
+    typename Group, typename S, typename T, size_t M, size_t N,
+    access::address_space Space, access::decorated IsDecorated,
+    typename = std::enable_if_t<std::is_same_v<S, std::remove_const_t<T>>>>
+void load_accumulator_hip(
+    joint_matrix_hip<
+        S, sycl::ext::oneapi::experimental::matrix::use::accumulator, M, N,
+        sycl::ext::oneapi::experimental::matrix::layout::dynamic> &res,
+    multi_ptr<T, Space, IsDecorated> src, size_t stride,
+    sycl::ext::oneapi::experimental::matrix::layout layout, Group &sg) {
+  static_assert(std::is_same_v<S, int32_t> || std::is_same_v<S, float> ||
```

Comment:
I don't think this assert can ever be false because such a joint_matrix shouldn't be constructable. Generally I think it is better to just make unsupported joint_matrix structs non-constructable, then you generally don't have to worry about such asserts in any function that takes joint_matrix as an argument. I think this is already the case in your impl.

Reply:
they are removed. @JackAKirk